oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/ServletBearerExchangeFilterFunction.java

```diff
@@ -0,0 +1,248 @@
+ *
+ * Suitable for Servlet applications, applying it to a typical {@link org.springframework.web.reactive.function.client.WebClient}
+ * configuration:
+ *
+ * <pre>
+ *  @Bean
+ *  WebClient webClient() {
+ *      ServletBearerExchangeFilterFunction bearer = new ServletBearerExchangeFilterFunction();
+ *      return WebClient.builder()
+ *              .apply(bearer.oauth2Configuration())
+ *              .build();
+ *  }
+ * </pre>
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ */
+public class ServletBearerExchangeFilterFunction
+		implements ExchangeFilterFunction, InitializingBean, DisposableBean {
+
+	private static final String AUTHENTICATION_ATTR_NAME = Authentication.class.getName();
+
+	private static final String REQUEST_CONTEXT_OPERATOR_KEY = RequestContextSubscriber.class.getName();
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		Hooks.onLastOperator(REQUEST_CONTEXT_OPERATOR_KEY,
+				Operators.liftPublisher((s, sub) -> createRequestContextSubscriber(sub)));
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void destroy() throws Exception {
+		Hooks.resetOnLastOperator(REQUEST_CONTEXT_OPERATOR_KEY);
+	}
+
+	/**
+	 * Configures the builder with {@link #defaultRequest()} and adds this as a {@link ExchangeFilterFunction}
+	 * @return the {@link Consumer} to configure the builder
+	 */
+	public Consumer<WebClient.Builder> oauth2Configuration() {
+		return builder -> builder.defaultRequest(defaultRequest()).filter(this);
+	}
+
+	/**
+	 * Provides defaults for the {@link Authentication} using
+	 * {@link SecurityContextHolder}. It also can default the {@link AbstractOAuth2Token} using the
+	 * {@link #authentication(Authentication)}.
+	 * @return the {@link Consumer} to populate the attributes
+	 */
+	public Consumer<WebClient.RequestHeadersSpec<?>> defaultRequest() {
+		return spec -> spec.attributes(attrs -> {
+			populateDefaultAuthentication(attrs);
+		});
+	}
+
+	/**
+	 * Modifies the {@link ClientRequest#attributes()} to include the {@link Authentication} used to
+	 * look up and save the {@link AbstractOAuth2Token}. The value is defaulted in
+	 * {@link ServletBearerExchangeFilterFunction#defaultRequest()}
+	 *
+	 * @param authentication the {@link Authentication} to use.
+	 * @return the {@link Consumer} to populate the attributes
+	 */
+	public static Consumer<Map<String, Object>> authentication(Authentication authentication) {
+		return attributes -> attributes.put(AUTHENTICATION_ATTR_NAME, authentication);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public Mono<ClientResponse> filter(ClientRequest request, ExchangeFunction next) {
+		return mergeRequestAttributesIfNecessary(request)
+				.filter(req -> req.attribute(AUTHENTICATION_ATTR_NAME).isPresent())
+				.map(req -> getOAuth2Token(req.attributes()))
+				.map(token -> bearer(request, token))
+				.flatMap(next::exchange)
+				.switchIfEmpty(Mono.defer(() -> next.exchange(request)));
```

Comment:
I know there are activities to reduce number of `Optional` usage and replacement of `filter/map/flatMap()` to the `handle()` in the spring core  
so it could be rewritten to  
may be it is a little bit more imperative, but it creates less operators and does not use create `Optional` to check attribute presence 
```java
public Mono<ClientResponse> filter(ClientRequest request, ExchangeFunction next) {
	return mergeRequestAttributesIfNecessary(request)
			.handle((req, sink) -> {
				Map<String, Object> attributes = req.attributes();
				if (attributes.get(AUTHENTICATION_ATTR_NAME) == null) {
					sink.next(request);
				} else {
					AbstractOAuth2Token token = getOAuth2Token(attributes);
					sink.next(bearer(req, token));
				}
			})
			.flatMap(next::exchange);
}

private Mono<ClientRequest> mergeRequestAttributesIfNecessary(ClientRequest request) {
	if (request.attributes().get(AUTHENTICATION_ATTR_NAME) != null) {
		return Mono.just(request);
	}
	return mergeRequestAttributesFromContext(request);
}

private Mono<ClientRequest> mergeRequestAttributesFromContext(ClientRequest request) {
	return Mono.subscriberContext()
			.map(ctx -> ClientRequest.from(request)
					.attributes(attrs -> populateRequestAttributes(attrs, ctx))
					.build()
			);
}
```